### DISCUSSION TITLE
Short-circuit evaluation for rules?

### DISCUSSION THREAD

**Discussion by devent** [evrete/evrete#106] [General]
Hi. If I have a where clause like this **(A)** `@Where(value={ "$f.x < $f.w1", "$f.z > 0", "$f.isNeighborsFilled(U)", "!$f.isNeighborsFilled(E)" })`  then each test is evaluated, even though they are connected with a logical AND. There is no Short-circuit evaluation done for all the tests. Meaning all tests are run even though $f.z = 0 and $f.z > 0 always returns false, so no other test needs to be done.

Can Short-circuit evaluation be activated?

I could write the where clause like this **(B)** `@Where(value={ "$f.x < $f.w1 && $f.z > 0 && $f.isNeighborsFilled(U) && !$f.isNeighborsFilled(E)" })`

I don't know if it is better to write each test individually **(A)** or write the test as a one big test with AND **(B)**. A benchmark shows that one big test with AND **(B)** is faster.

**Comment by andbi** [comment]
Hi!

When it comes to alpha conditions (those involving one fact type), the results of their evaluation form a bitmask, which in turn is a memory bucket address. Thus, this process cannot be avoided.

However, once an alpha condition is evaluated, the same condition in other rules will reuse the previously obtained result. So, if the conditions in your example are never reused, I would opt for the **B** variant. Or, even faster, you could declare a custom boolean field.

As for the beta conditions (involving multiple fact types), the engine will short-circuit the evaluation phase.

**Comment by devent** [comment]
> the same condition in other rules will reuse the previously obtained result. So, if the conditions in your example are never reused, I would opt for the B variant.

Ok this makes sense. Thx for the reply.
